### PR TITLE
Load params from registers when possible

### PR DESCRIPTION
**Pull request by LouisLaf**
Load ArgIns from registers if the param was passed in registers (x64 only).  Cleanup implicitParams in the process.

### CODE REVIEW

**Comment by LouisLaf** [comment]
@satheeshravi, @rajatd, please review.

File: lib/Backend/Sym.h [hunk 1]
```diff
@@ -162,11 +164,11 @@ class StackSym: public Sym
     bool            IsTempReg(Func *const func) const;
     bool            IsFromByteCodeConstantTable() const { return m_isFromByteCodeConstantTable; }
     void            SetIsFromByteCodeConstantTable() { this->m_isFromByteCodeConstantTable = true; }
-    Js::ArgSlot     GetArgSlotNum() const { Assert(HasArgSlotNum()); return m_argSlotNum; }
-    bool            HasArgSlotNum() const { return m_argSlotNum != 0; }
+    Js::ArgSlot     GetArgSlotNum() const { Assert(HasArgSlotNum()); return m_slotNum; }
+    bool            HasArgSlotNum() const { return !!(m_isArgSlotSym | m_isArgSlotRegSym); }
```
Comment (satheeshravi) [inline lib/Backend/Sym.h]: > HasArgSlotNum() [](start = 19, length = 16)

Do you also want to assert that m_slotNum has a slot number, when either of the bool is true ?

File: lib/Backend/Sym.inl [hunk 2]
```diff
@@ -86,7 +86,7 @@ StackSym::IsArgSlotSym() const
 inline bool
 StackSym::IsParamSlotSym() const
 {
-    return m_paramSlotNum != InvalidSlot;
+	return m_isParamSym;
```
Comment (satheeshravi) [inline lib/Backend/Sym.inl]: >return m_isParamSym; [](start = 1, length = 20)

Assert for slotNum here ?

File: lib/Backend/SymTable.h [hunk 3]
```diff
@@ -9,21 +9,25 @@ class SymTable
...
     typedef JsUtil::BaseDictionary<Js::PropertyId, BVSparse<JitArenaAllocator> *, JitArenaAllocator, PowerOf2SizePolicy> PropertyEquivBvMap;
-public:
-    static const int    k_symTableSize = 8192;
-    Sym *               m_table[k_symTableSize];
-    PropertyMap *       m_propertyMap;
-    PropertyEquivBvMap *m_propertyEquivBvMap;
 
 private:
+	static const int    k_symTableSize = 8192;
+	static const int    k_maxImplicitParamSlot = 4;
```
Comment (satheeshravi) [inline lib/Backend/SymTable.h]: >k_maxImplicitParamSlot [](start = 21, length = 22)

just curious... why is it k_ ? :)

File: lib/Backend/amd64/LinearScanMD.cpp [hunk 4]
```diff
@@ -450,3 +450,93 @@ RegNum LinearScanMD::GetRegisterFromSaveIndex(uint offset)
...
+
+	if (func->GetJnFunction()->GetIsAsmjsMode() && !func->IsLoopBody())
+	{
+		// Asm.js function only have 1 implicit param as they have no CallInfo, and they have float/SIMD params.
+		// Asm.js loop bodies however are called like normal JS functions.
+		if (IRType_IsFloat(symOpnd->GetType()) || IRType_IsSimd(symOpnd->GetType()))
+		{
+			switch (paramSym->GetParamSlotNum())
+			{
+			case 10000:
```
Comment (satheeshravi) [inline lib/Backend/amd64/LinearScanMD.cpp]: >10000 [](start = 8, length = 5)

How can slot num be this value ?

File: lib/Backend/Sym.cpp [hunk 5]
```diff
@@ -109,20 +112,24 @@ StackSym::New(IRType type, Func *func)
...
 StackSym::NewParamSlotSym(Js::ArgSlot paramSlotNum, Func * func)
 {
-    StackSym * stackSym = StackSym::New(func);
-    stackSym->m_paramSlotNum = paramSlotNum;
-    return stackSym;
+	return NewParamSlotSym(paramSlotNum, func, TyVar);
 }
 
 StackSym *
 StackSym::NewParamSlotSym(Js::ArgSlot paramSlotNum, Func * func, IRType type)
```
Comment (satheeshravi) [inline lib/Backend/Sym.cpp]: > IRType type [](start = 64, length = 12)

nit: you can consider adding a default value for this parameter, and removing the other definition for NewParamSlotSym

File: lib/Backend/Sym.cpp [hunk 6]
```diff
@@ -987,117 +996,133 @@ Sym::Dump(IRDumpFlags flags, const ValueType valueType)
...
+					Output::Print(_u("callInfo"));
+					break;
+				case 2:
+					Output::Print(_u("funcInfo"));
+					break;
+				case 3:
+					Output::Print(_u("genObj"));
+					break;
+				case 4:
+					Output::Print(_u("genFrame"));
```
Comment (satheeshravi) [inline lib/Backend/Sym.cpp]: >genFrame [](start = 23, length = 8)

Should be ResumeYieldData?

File: lib/Backend/SymTable.cpp [hunk 7]
```diff
@@ -243,6 +243,23 @@ SymTable::GetArgSlotSym(Js::ArgSlot argSlotNum)
     return argSym;
 }
 
+StackSym *          
+SymTable::GetImplicitParam(Js::ArgSlot paramSlotNum)
+{
+	Assert(paramSlotNum - 1 <= this->k_maxImplicitParamSlot);
```
Comment (satheeshravi) [inline lib/Backend/SymTable.cpp]: ><= [](start = 25, length = 2)

just "<" ?

**Comment by satheeshravi** [comment]
> }

I think, this function is unused. Can you please remove ?

---
Refers to: lib/Backend/LowerMDShared.cpp:8044 in cc151bf. [](commit_id = cc151bf0953a34d459c8a380ca65e31d8d53b0d6, deletion_comment = True)

File: lib/Backend/Sym.cpp [hunk 8]
```diff
@@ -987,117 +996,133 @@ Sym::Dump(IRDumpFlags flags, const ValueType valueType)
...
+			}
+			Output::Print(_u("(s%d)"), m_id);
+		}
+		else if (stackSym->IsParamSlotSym())
+		{
+			if (stackSym->IsImplicitParamSym())
+			{
+				switch (stackSym->GetParamSlotNum())
+				{
+				case 1:
```
Comment (satheeshravi) [inline lib/Backend/Sym.cpp]: >case 1 [](start = 4, length = 6)

It looks like, the index starts from 0 , and 0 is usually is for functionObject. It looks to me like slotNum is  not strictly compliant with argOffset.

File: lib/Backend/SymTable.h [hunk 9]
```diff
@@ -9,21 +9,25 @@ class SymTable
...
+    PropertyMap *       m_propertyMap;
     Func *              m_func;
     SymID               m_currentID;
     SymID               m_IDAdjustment;
 
 public:
+	PropertyEquivBvMap *m_propertyEquivBvMap;
+
+public:
     SymTable() : m_currentID(0), m_func(nullptr), m_IDAdjustment(0)
```
Comment (satheeshravi) [inline lib/Backend/SymTable.h]: > SymTable() [](start = 3, length = 11)

Initialize m_propertyMap, m_propertyEquivBvMap ?

Comment (satheeshravi) [inline lib/Backend/SymTable.h]: just noticed that this is initialized in Init(). So this is a nit.

---
In reply to: [75063468](https://github.com/Microsoft/ChakraCore/pull/1373#discussion_r75063468) [](ancestors = 75063468)

File: lib/Backend/amd64/LinearScanMD.cpp [hunk 10]
```diff
@@ -450,3 +450,93 @@ RegNum LinearScanMD::GetRegisterFromSaveIndex(uint offset)
...
+			}
+		}
+	}
+	else // Non-Asm.js
+	{
+		Assert(symOpnd->GetType() == TyVar || IRType_IsNativeInt(symOpnd->GetType()));
+
+		if (paramSym->IsImplicitParamSym())
+		{
+			switch (paramSym->GetParamSlotNum())
```
Comment (satheeshravi) [inline lib/Backend/amd64/LinearScanMD.cpp]: >switch (paramSym->GetParamSlotNum()) [](start = 3, length = 36)

How will this work for generators which has 3rd and 4th slots ?

File: lib/Backend/SymTable.h (hunk 3 above)
Comment (LouisLaf) [inline lib/Backend/SymTable.h]: k_ is for constants.  Originates from Hungarian notations.

---
In reply to: [73779694](https://github.com/Microsoft/ChakraCore/pull/1373#discussion_r73779694) [](ancestors = 73779694)

File: lib/Backend/amd64/LinearScanMD.cpp (hunk 10 above)
Comment (LouisLaf) [inline lib/Backend/amd64/LinearScanMD.cpp]: They just get loaded from the stack for now.  We can fix this when optimizations are enabled for generators.

---
In reply to: [75064004](https://github.com/Microsoft/ChakraCore/pull/1373#discussion_r75064004) [](ancestors = 75064004)

**Comment by LouisLaf** [comment]
> }

It is used by LoadCallInfo in lower.

---
In reply to: [240313560](https://github.com/Microsoft/ChakraCore/pull/1373#issuecomment-240313560) [](ancestors = 240313560)

---
Refers to: lib/Backend/LowerMDShared.cpp:8044 in cc151bf. [](commit_id = cc151bf0953a34d459c8a380ca65e31d8d53b0d6, deletion_comment = True)

File: lib/Backend/Sym.cpp (hunk 8 above)
Comment (LouisLaf) [inline lib/Backend/Sym.cpp]: 0 is not valid.

---
In reply to: [75063249](https://github.com/Microsoft/ChakraCore/pull/1373#discussion_r75063249) [](ancestors = 75063249)

File: lib/Backend/Sym.cpp (hunk 6 above)
Comment (LouisLaf) [inline lib/Backend/Sym.cpp]: Same thing as the generator frame.

---
In reply to: [75061008](https://github.com/Microsoft/ChakraCore/pull/1373#discussion_r75061008) [](ancestors = 75061008)

**Comment by satheeshravi** [comment]
:shipit:

**Comment by obastemur** [comment]
It could be really nice to squash all these 'cleanup' and 'more cleanup' commits before merging.

**Pull request merged**